src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java

```diff
@@ -0,0 +1,426 @@
+    if (conn == null || conn.isClosed()) {
+      throw SnowflakeErrors.ERROR_5010.getException();
+    }
+
+    this.fileSize = SnowflakeSinkConnectorConfig.BUFFER_SIZE_BYTES_DEFAULT;
+    this.recordNum = SnowflakeSinkConnectorConfig.BUFFER_COUNT_RECORDS_DEFAULT;
+    this.flushTime = SnowflakeSinkConnectorConfig.BUFFER_FLUSH_TIME_SEC_DEFAULT;
+    this.conn = conn;
+    this.recordService = new RecordService();
+    this.telemetryService = conn.getTelemetryClient();
+    this.topic2TableMap = new HashMap<>();
+
+    // Setting the default value in constructor
+    // meaning it will not ignore the null values (Tombstone records wont be ignored/filtered)
+    this.behaviorOnNullValues = SnowflakeSinkConnectorConfig.BehaviorOnNullValues.DEFAULT;
+
+    this.connectorConfig = connectorConfig;
+    // Streaming ingest requires a properties.
+    // TODO request a builder pattern for SI Client.
+    Properties streamingClientProps = new Properties();
+    streamingClientProps.putAll(connectorConfig);
+    this.taskId = connectorConfig.getOrDefault(Utils.TASK_ID, "-1");
+    this.streamingIngestClientName =
+        STREAMING_CLIENT_PREFIX_NAME + conn.getConnectorName() + "_" + taskId;
+    this.streamingIngestClient =
+        SnowflakeStreamingIngestClientFactory.builder(this.streamingIngestClientName)
+            .setProperties(streamingClientProps)
+            .build();
+    this.partitionsToChannel = new HashMap<>();
+  }
+
+  @Override
+  public void startTask(String tableName, String topic, int partition) {
+    String partitionChannelKey = partitionChannelKey(topic, partition);
+    LOGGER.info(
+        "Opening a channel with name:{} for table name:{}, topic:{}, partition:{}",
+        partitionChannelKey,
+        tableName,
+        topic,
+        partition);
+    // the table should be present before opening a channel so lets do a table existence check here
+    createTableIfNotExists(tableName);
+    SnowflakeStreamingIngestChannel partitionChannel =
+        streamingIngestClient.openChannel(getOpenChannelRequest(partitionChannelKey, tableName));
+    TopicPartitionChannel topicPartitionChannel =
+        new TopicPartitionChannel(partitionChannel, this.conn, tableName);
+    partitionsToChannel.putIfAbsent(partitionChannelKey, topicPartitionChannel);
+  }
+
+  /**
+   * Inserts the given record into buffer and then eventually calls insertRows API if buffer
+   * threshold has reached.
+   *
+   * <p>TODO: SNOW-473896 - Please note we will get away with Buffering logic in future commits.
+   *
+   * @param records record content
+   */
+  @Override
+  public void insert(Collection<SinkRecord> records) {
+    // note that records can be empty
+    for (SinkRecord record : records) {
+      // check if need to handle null value records
+      if (recordService.shouldSkipNullValue(record, behaviorOnNullValues)) {
+        continue;
+      }
+      // Might happen a count of record based flushing
+      insert(record);
```

Comment:
What will happen if the insertion failed? I guess we have roughly two cases, if the error is channel related, we would want to reopen the channel and start with the last offset token. If the error is record related, we would want to insert it to the DLQ?